src/main/kotlin/com/openlattice/mechanic/upgrades/SyncOrgPermissionsUpgrade.kt

```diff
@@ -0,0 +1,143 @@
+
+import com.hazelcast.query.Predicates
+import com.openlattice.authorization.Ace
+import com.openlattice.authorization.AceKey
+import com.openlattice.authorization.AceValue
+import com.openlattice.authorization.Acl
+import com.openlattice.authorization.AclKey
+import com.openlattice.authorization.Action
+import com.openlattice.authorization.DbCredentialService
+import com.openlattice.authorization.PrincipalType
+import com.openlattice.authorization.SecurablePrincipal
+import com.openlattice.authorization.mapstores.PermissionMapstore
+import com.openlattice.authorization.mapstores.PrincipalMapstore
+import com.openlattice.authorization.securable.SecurableObjectType
+import com.openlattice.edm.PropertyTypeIdFqn
+import com.openlattice.edm.set.EntitySetFlag
+import com.openlattice.hazelcast.HazelcastMap
+import com.openlattice.mechanic.Toolbox
+import com.openlattice.postgres.external.ExternalDatabaseConnectionManager
+import com.openlattice.postgres.external.ExternalDatabasePermissioningService
+import com.openlattice.transporter.processors.GetPropertyTypesFromTransporterColumnSetEntryProcessor
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+/**
+ * @author Drew Bailey (drew@openlattice.com)
+ */
+class SyncOrgPermissionsUpgrade(
+        toolbox: Toolbox,
+        private val exConnMan: ExternalDatabaseConnectionManager,
+        private val exDbPermMan: ExternalDatabasePermissioningService,
+        private val dbCreds: DbCredentialService
+): Upgrade {
+
+    val logger: Logger = LoggerFactory.getLogger(SyncOrgPermissionsUpgrade::class.java)
+
+    private val entitySets = toolbox.entitySets
+    private val propertyTypes = toolbox.propertyTypes
+    private val transporterState = HazelcastMap.TRANSPORTER_DB_COLUMNS.getMap(toolbox.hazelcast)
+    private val principalTrees = HazelcastMap.PRINCIPAL_TREES.getMap(toolbox.hazelcast)
+    private val princpals = HazelcastMap.PRINCIPALS.getMap(toolbox.hazelcast)
+    private val permissions = HazelcastMap.PERMISSIONS.getMap(toolbox.hazelcast)
+
+    override fun upgrade(): Boolean {
+        val addRolesToDbCreds = addRolesToDbCreds()
+        val assemblies = initializeAssemblyPermissions()
+        val mapPTrees = mapAllPrincipalTrees()
+        val createPRoles = createAssignAllPermRoles()
+        if (addRolesToDbCreds && assemblies  && mapPTrees  && createPRoles) {
+            return true
+        }
+        logger.error("Sync permissions upgrade failed, final status:\n" +
+                "addRolesToDbCreds: {}\n" +
+                "updatePermissionsForAssemblies: {}\n" +
+                "mapAllPrincipalTrees: {}\n" +
+                "createAllPermRoles: {}\n", addRolesToDbCreds, assemblies, mapPTrees, createPRoles)
+        return false
+    }
+
+    private fun addRolesToDbCreds(): Boolean {
+        princpals.values(
+                Predicates.equal<AclKey, SecurablePrincipal>(PrincipalMapstore.PRINCIPAL_TYPE_INDEX, PrincipalType.ROLE)
+        ).map { rolePrincipal ->
```

Comment:
might want to do a `forEach` instead since it's not mapping to anything and I've definitely run into issues in the past where a `.map` block ends up not getting executed at all if it's not terminating to a collection or something

Reply:
Shouldn't happen in Kotlin b/c `.map{}` is terminal but I can change it